lib/booking_manager.rb

```diff
@@ -0,0 +1,125 @@
+require_relative 'room'
+require_relative 'single_res'
+require_relative 'block_res'
+require_relative 'date_range'
+
+module Hotel
+  class BookingManager
+    attr_reader :rooms, :reservations, :blocks
+    
+    def initialize(total_room_count)
+      @rooms = Hotel::Room.all_rooms(total_room_count)
+      @reservations = []
+      @blocks = []
+    end
+    
+    def list_all_rooms
+      room_list = @rooms.map do |room|
+        "Room #{room.number}"
+      end.join("\n")
+      
+      return room_list
+    end
+    
+    def find_reservations_by_date(date)
+      found_singles = reservations.select do |reservation|
+        reservation.date_range.date_included?(date)
+      end
+
+      found_blocks = blocks.select do |block|
+        block.date_range.date_included?(date)
+      end
+
+      found_reservations = found_singles + found_blocks
+      
+      if found_reservations.length == 0
+        nil
+      else
+        return found_reservations
+      end 
+    end
+
+    def rooms_available(start_date, end_date)
+      reserved_single = reservations.select do |reservation|
+        reservation.date_range.overlaps?(start_date, end_date)
+      end.map! {|reservation| reservation.room}
+
+      reserved_blocks = blocks.select do |block|
+        block.date_range.overlaps?(start_date, end_date)
+      end.map! {|block| block.unreserved_rooms}
+
+      if reserved_single.length > 0 && reserved_blocks.length > 0
+        all_reserved = reserved_single + reserved_blocks.flatten!
+      elsif reserved_single.length > 0 && reserved_blocks.length == 0
+        all_reserved = reserved_single
+      elsif reserved_single.length == 0 && reserved_blocks.length > 0
+        all_reserved = reserved_blocks.flatten!
+      else
+        all_reserved = []
+      end
```

Comment:
This is really strange, but I think that if you had more time on this code, there would be a really interesting way to refactor this code so that this `if ... else` block didn't exist. For example, no matter what, assuming that `reserved_single` and `reserved_blocks` are empty arrays, then `all_reserved = reserved_single + reserved_blocks`. The struggle here is that `reserved_blocks` is either a nested array or `nil` at the current implementation. However, just as a starting point, consider that replacing this block of `if ... else` with this line makes your tests pass still:

```ruby
all_reserved = reserved_single + reserved_blocks.flatten!.to_a
```